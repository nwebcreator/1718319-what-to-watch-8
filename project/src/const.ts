export enum AuthorizationStatus {
  Auth = 'AUTH',
  NoAuth = 'NO_AUTH',
  Unknown = 'UNKNOWN',
}

export const FILMS_PER_STEP = 8;

export const ALL_GENRES_NAME = 'All genres';
