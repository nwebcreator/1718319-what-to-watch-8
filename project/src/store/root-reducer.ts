import { combineReducers } from 'redux';
import { filmsData } from './films-data/films-data';
import { userProcess } from './user-process/user-process';

export enum NameSpace {
  Data = 'DATA',
  User = 'USER',
}

export const rootReducer = combineReducers({
  [NameSpace.Data]: filmsData,
  [NameSpace.User]: userProcess,
});

export type RootState = ReturnType<typeof rootReducer>;
