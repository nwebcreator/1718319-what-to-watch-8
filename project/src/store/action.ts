import { AuthorizationStatus } from '../const';
import { AppRoute } from '../routes';
import { ActionType } from '../types/action';
import { Film, Films } from '../types/film';
import { AuthInfo } from '../types/auth-info';
import { Reviews } from '../types/review';
import { createAction } from '@reduxjs/toolkit';

export const changeGenre = createAction(
  ActionType.ChangeGenre,
  (genre: string) => ({
    payload: genre,
  }),
);

export const loadFilms = createAction(
  ActionType.LoadFilms,
  (films: Films) => ({
    payload: {
      films,
    },
  }),
);

export const loadFavoriteFilms = createAction(
  ActionType.LoadFavoriteFilms,
  (favoriteFilms: Films) => ({
    payload: {
      favoriteFilms,
    },
  }),
);

export const updateFilmFavoriteStatus = createAction(
  ActionType.UpdateFilmFavoriteStatus,
  (id: number, isFavorite: boolean) => ({
    payload: {
      id,
      isFavorite,
    },
  }),
);

export const loadReviews = createAction(
  ActionType.LoadReviews,
  (reviews: Reviews) => ({
    payload: {
      reviews,
    },
  }),
);

export const loadSimilarFilms = createAction(
  ActionType.LoadSimilarFilms,
  (similarFilms: Films) => ({
    payload: {
      similarFilms,
    },
  }),
);

export const loadPromoFilm = createAction(
  ActionType.LoadPromoFilm,
  (promoFilm: Film) => ({
    payload: {
      promoFilm,
    },
  }),
);

export const loadFilm = createAction(
  ActionType.LoadFilm,
  (film: Film) => ({
    payload: {
      film,
    },
  }),
);

export const changeShowedFilms = createAction(
  ActionType.ChangeShowedFilms,
  (showedFilms: number) => ({
    payload: showedFilms,
  }),
);

export const requireAuthorization = createAction(
  ActionType.RequireAuthorization,
  (authorizationStatus: AuthorizationStatus, authInfo?: AuthInfo) => ({
    payload: {
      authorizationStatus,
      authInfo,
    },
  }),
);

export const requireLogout = createAction(ActionType.RequireLogout);

export const redirectToRoute = createAction(
  ActionType.RedirectToRoute,
  (url: AppRoute | string) => ({
    payload: url,
  }),
);
