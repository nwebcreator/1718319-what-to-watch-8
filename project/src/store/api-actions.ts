import { ThunkActionResult } from '../types/action';
import { loadFavoriteFilms, loadFilm, loadFilms, loadPromoFilm, loadReviews, loadSimilarFilms, redirectToRoute, requireAuthorization, requireLogout, updateFilmFavoriteStatus } from './action';
import { saveToken, dropToken } from '../services/token';
import { AuthorizationStatus } from '../const';
import { AuthData } from '../types/auth-data';
import { AppRoute } from '../routes';
import { mapDataToFilm } from '../mappers/film-mapper';
import { mapDataToAuthInfo } from '../mappers/auth-info-mapper';
import { Reviews } from '../types/review';
import { AddReview } from '../types/add-review';
import { toast } from 'react-toastify';

const FETCH_FAIL_MESSAGE = 'Ошибка загрузки данных с сервера. Попробуйте позже.';
const NO_AUTH_FAIL_MESSAGE = 'Не забудьте авторизоваться';
const AUTH_FAIL_MESSAGE = 'При авторизации возникла ошибка. Попробуйте позже.';
const LOGOUT_FAIL_MESSAGE = 'При выходе из системы возникла ошибка. Попробуйте позже.';

type ResponseValueType = string | string[] | number | boolean;
export type ApiResponse = Record<string, ResponseValueType | Record<string, ResponseValueType>>;

export const fetchPromoAction = (): ThunkActionResult =>
  async (dispatch, _getState, api): Promise<void> => {
    try {
      const { data } = await api.get<ApiResponse>('/promo');
      dispatch(loadPromoFilm(mapDataToFilm(data)));
    } catch {
      toast.error(FETCH_FAIL_MESSAGE);
    }
  };

export const fetchFilmsAction = (): ThunkActionResult =>
  async (dispatch, _getState, api): Promise<void> => {
    try {
      const { data } = await api.get<ApiResponse[]>('/films');
      dispatch(loadFilms(data.map((it) => mapDataToFilm(it))));
    } catch {
      toast.error(FETCH_FAIL_MESSAGE);
    }
  };

export const fetchFavoriteFilmsAction = (): ThunkActionResult =>
  async (dispatch, _getState, api): Promise<void> => {
    try {
      const { data } = await api.get<ApiResponse[]>('/favorite');
      dispatch(loadFavoriteFilms(data.map((it) => mapDataToFilm(it))));
    } catch {
      toast.error(FETCH_FAIL_MESSAGE);
    }
  };

export const fethcSimilarFilmsAction = (id: number): ThunkActionResult =>
  async (dispatch, _getState, api): Promise<void> => {
    try {
      const { data } = await api.get<ApiResponse[]>(`/films/${id}/similar`);
      dispatch(loadSimilarFilms(data.map((it) => mapDataToFilm(it)).filter((it) => it.id !== id)));
    } catch {
      toast.error(FETCH_FAIL_MESSAGE);
    }
  };

export const fetchReviewsAction = (id: number): ThunkActionResult =>
  async (dispatch, _getSetate, api): Promise<void> => {
    try {
      const { data } = await api.get<Reviews>(`/comments/${id}`);
      dispatch(loadReviews(data));
    }
    catch {
      toast.error(FETCH_FAIL_MESSAGE);
    }
  };

export const fetchFilmAction = (id: number): ThunkActionResult =>
  async (dispatch, _getState, api): Promise<void> => {
    try {
      const { data } = await api.get<ApiResponse>(`/films/${id}`);
      dispatch(loadFilm(mapDataToFilm(data)));
    } catch {
      toast.error(FETCH_FAIL_MESSAGE);
    }
  };

export const addReviewAction = (id: number, review: AddReview, cb: (isSuccess: boolean) => void): ThunkActionResult =>
  async (dispatch, _getState, api): Promise<void> => {
    try {
      const { data } = await api.post<Reviews>(`/comments/${id}`, review);
      dispatch(loadReviews(data));
      toast.info('Отзыв успешно добавлен');
      cb(true);
    } catch {
      toast.error('Ошибка при отправке отзыва.');
      cb(false);
    }
  };

export const changeFilmFavoriteStatus = (id: number, isFavorite: boolean): ThunkActionResult =>
  async (dispatch, _getState, api): Promise<void> => {
    try {
      const { data } = await api.post<ApiResponse>(`/favorite/${id}/${isFavorite ? 1 : 0}`);
      const film = mapDataToFilm(data);
      dispatch(updateFilmFavoriteStatus(film.id, film.isFavorite));
    } catch {
      toast.error(FETCH_FAIL_MESSAGE);
    }
  };

export const checkAuthAction = (): ThunkActionResult =>
  async (dispatch, _getState, api) => {
    try {
      const { data } = await api.get<ApiResponse>('/login');
      const authInfo = mapDataToAuthInfo(data);
      const authorizationStatus = authInfo ? AuthorizationStatus.Auth : AuthorizationStatus.NoAuth;
      dispatch(requireAuthorization(authorizationStatus, authInfo));
      if (authorizationStatus !== AuthorizationStatus.Auth) {
        toast.info(NO_AUTH_FAIL_MESSAGE);
      }
    } catch {
      toast.info(NO_AUTH_FAIL_MESSAGE);
    }
  };

export const loginAction = ({ email, password }: AuthData): ThunkActionResult =>
  async (dispatch, _getState, api) => {
    try {
      const { data } = await api.post<ApiResponse>('/login', { email, password });
      const authInfo = mapDataToAuthInfo(data);
      if (authInfo) {
        saveToken(authInfo.token);
        dispatch(requireAuthorization(AuthorizationStatus.Auth, authInfo));
      }

      dispatch(redirectToRoute(AppRoute.Root));
    } catch {
      toast.info(AUTH_FAIL_MESSAGE);
    }
  };

export const logoutAction = (): ThunkActionResult =>
  async (dispatch, _getState, api) => {
    try {
      await api.delete('/logout');
      dropToken();
      dispatch(requireLogout());
    } catch {
      toast.error(LOGOUT_FAIL_MESSAGE);
    }
  };
