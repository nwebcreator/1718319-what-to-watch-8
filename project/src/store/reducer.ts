import { createReducer } from '@reduxjs/toolkit';
import { ALL_GENRES_NAME, AuthorizationStatus } from '../const';
import { State } from '../types/state';
import { changeGenre, changeShowedFilms, loadFavoriteFilms, loadFilm, loadFilms, loadReviews, loadSimilarFilms, requireAuthorization, requireLogout } from './action';

const initialState: State = {
  activeGenre: ALL_GENRES_NAME,
  films: [],
  favoriteFilms: [],
  reviews: [],
  similarFilms: [],
  film: undefined,
  showedFilms: 0,
  authorizationStatus: AuthorizationStatus.Unknown,
  authInfo: undefined,
  isDataLoaded: false,
};

const reducer = createReducer(initialState, (builder) => {
  builder
    .addCase(changeGenre, (state, action) => {
      state.activeGenre = action.payload;
    })
    .addCase(loadFilms, (state, action) => {
      state.films = action.payload.films;
      state.isDataLoaded = true;
    })
    .addCase(loadFavoriteFilms, (state, action) => {
      state.favoriteFilms = action.payload.films;
    })
    .addCase(loadSimilarFilms, (state, action) => {
      state.similarFilms = action.payload.similarFilms;
    })
    .addCase(loadReviews, (state, action) => {
      state.reviews = action.payload.reviews;
    })
    .addCase(loadFilm, (state, action) => {
      state.film = action.payload.film;
    })
    .addCase(requireAuthorization, (state, action) => {
      state.authorizationStatus = action.payload.authorizationStatus;
      state.authInfo = action.payload.authInfo;
    })
    .addCase(requireLogout, (state) => {
      state.authorizationStatus = AuthorizationStatus.NoAuth;
      state.authInfo = undefined;
    })
    .addCase(changeShowedFilms, (state, acion) => {
      state.showedFilms = acion.payload;
    });
});

export { reducer };
