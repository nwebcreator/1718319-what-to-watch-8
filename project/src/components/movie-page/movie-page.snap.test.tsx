import { configureMockStore } from '@jedmao/redux-mock-store';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { AuthorizationStatus } from '../../const';
import { NameSpace } from '../../store/root-reducer';
import MoviePage from './movie-page';

const mockStore = configureMockStore();
const history = createMemoryHistory();

describe('Component: MoviePage', () => {
  test('should render correctly', () => {
    const film = {
      id: 1,
      name: 'string',
      posterImage: 'string',
      previewImage: 'string',
      backgroundImage: 'string',
      backgroundColor: 'string',
      videoLink: 'string',
      previewVideoLink: 'string',
      description: 'string',
      rating: 10,
      scoresCount: 10,
      director: 'string',
      starring: [],
      runTime: 10,
      genre: 'string',
      released: 2010,
      isFavorite: false,
    };

    const store = mockStore({
      [NameSpace.User]: { authorizationStatus: AuthorizationStatus.Auth },
      [NameSpace.Data]: { film, similarFilms: [film] },
    });

    const { container } = render(
      <Provider store={store}>
        <Router history={history}>
          <MoviePage />
        </Router>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
