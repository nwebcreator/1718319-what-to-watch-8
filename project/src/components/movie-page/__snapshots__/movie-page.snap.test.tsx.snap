// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MoviePage should render correctly 1`] = `
<div>
  <section
    class="film-card film-card--full"
  >
    <div
      class="film-card__hero"
    >
      <div
        class="film-card__bg"
      >
        <img
          alt="string"
          src="string"
        />
      </div>
      <h1
        class="visually-hidden"
      >
        WTW
      </h1>
      <header
        class="page-header film-card__head"
      >
        <div
          class="logo"
        >
          <a
            class="logo__link"
            href="/"
          >
            <span
              class="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              class="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              class="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <ul
          class="user-block"
        >
          <li
            class="user-block__item"
          >
            <div
              class="user-block__avatar"
            >
              <a
                href="/mylist"
              >
                <img
                  alt="User avatar"
                  height="63"
                  src="img/avatar.jpg"
                  width="63"
                />
              </a>
            </div>
          </li>
          <li
            class="user-block__item"
          >
            <button
              class="user-block__link"
              style="background: none;"
            >
              Sign out
            </button>
          </li>
        </ul>
      </header>
      <div
        class="film-card__wrap"
      >
        <div
          class="film-card__desc"
        >
          <h2
            class="film-card__title"
          >
            string
          </h2>
          <p
            class="film-card__meta"
          >
            <span
              class="film-card__genre"
            >
              string
            </span>
            <span
              class="film-card__year"
            >
              2010
            </span>
          </p>
          <div
            class="film-card__buttons"
          >
            <button
              class="btn btn--play film-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlink:href="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              class="btn btn--list film-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlink:href="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              class="btn film-card__button"
              href="/films/NaN/review"
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="film-card__wrap film-card__translate-top"
    >
      <div
        class="film-card__info"
      >
        <div
          class="film-card__poster film-card__poster--big"
        >
          <img
            alt="string"
            height="327"
            src="string"
            width="218"
          />
        </div>
        <div
          class="film-card__desc"
        >
          <nav
            class="film-nav film-card__nav"
          >
            <ul
              class="film-nav__list"
            >
              <li
                class="film-nav__item film-nav__item--active"
              >
                <a
                  class="film-nav__link"
                  href="/#overview"
                >
                  Overview
                </a>
              </li>
              <li
                class="film-nav__item"
              >
                <a
                  class="film-nav__link"
                  href="/#details"
                >
                  Details
                </a>
              </li>
              <li
                class="film-nav__item"
              >
                <a
                  class="film-nav__link"
                  href="/#reviews"
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div
            class="film-rating"
          >
            <div
              class="film-rating__score"
            >
              10
            </div>
            <p
              class="film-rating__meta"
            >
              <span
                class="film-rating__level"
              >
                Awesome
              </span>
              <span
                class="film-rating__count"
              >
                10
                 ratings
              </span>
            </p>
          </div>
          <div
            class="film-card__text"
          >
            <p>
              string
            </p>
            <p
              class="film-card__director"
            >
              <strong>
                Director: 
                string
              </strong>
            </p>
            <p
              class="film-card__starring"
            >
              <strong>
                Starring: 
                
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    class="page-content"
  >
    <section
      class="catalog catalog--like-this"
    >
      <h2
        class="catalog__title"
      >
        More like this
      </h2>
      <div
        class="catalog__films-list"
      />
    </section>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link logo__link--light"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;
